File: samples js/LongMethodFeatureEnvy.js
class CodeSmellExamples {

    processOrder(order) {
        // Validation
        if (!order) {
            console.error("Order is null");
            return;
        }
        if (!order.items || order.items.length === 0) {
            console.error("Order has no items");
            return;
        }

        // Calculate totals
        let subtotal = 0;
        for (const item of order.items) {
            if (item.quantity <= 0) {
                console.error("Invalid quantity for item: " + item.productId);
                continue;
            }
            subtotal += item.price * item.quantity;
        }

        // Apply discounts
        let discount = 0;
        if (order.customer.premium) {
            discount = subtotal * 0.1;
        }
        if (subtotal > 100) {
            discount += 10.0;
        }

        // Calculate tax
        let taxRate = 0.08;
        const state = order.customer.address.state;
        if (state === "CA") {
            taxRate = 0.0725;
        } else if (state === "NY") {
            taxRate = 0.04;
        } else if (state === "TX") {
            taxRate = 0.0625;
        }
        const tax = (subtotal - discount) * taxRate;

        // Final total
        const total = subtotal - discount + tax;
        order.total = total;

        // Update inventory
        for (const item of order.items) {
            const inventory = this.getInventory();
            const product = inventory.findProduct(item.productId);
            if (product) {
                product.quantity -= item.quantity;
                inventory.updateProduct(product);
            }
        }

        // Save to database
        try {
            const db = this.getDatabase();
            db.save(order);
            console.log("Order saved: " + order.id);
        } catch (e) {
            console.error("Failed to save order: " + e.message);
            throw new Error("Order processing failed");
        }

        // Send confirmation
        const emailService = this.getEmailService();
        emailService.sendEmail(
            order.customer.email,
            "Order Confirmation",
            this.buildConfirmationEmail(order)
        );

        // Log activity
        console.log("Order processed: " + order.id);
        console.log("Customer: " + order.customer.firstName + " " + order.customer.lastName);
        console.log("Total: $" + total);
    }

    generateCustomerReport(customer) {
        let report = "";
        report += "Customer Report\n";
        report += "================\n\n";

        report += "Name: " + customer.firstName + " " + customer.lastName + "\n";
        report += "Email: " + customer.email + "\n";
        report += "Phone: " + customer.phone + "\n";

        const address = customer.address;
        report += "Address: " + address.street + ", ";
        report += address.city + ", ";
        report += address.state + " ";
        report += address.zipCode + "\n\n";

        report += "Account Status: ";
        if (customer.premium) {
            report += "Premium Member\n";
            report += "Member Since: " + customer.memberSince + "\n";
        } else {
            report += "Standard Member\n";
        }

        report += "\nOrder History:\n";
        let orderCount = 0;
        let totalSpent = 0;
        for (const order of customer.orders) {
            orderCount++;
            totalSpent += order.total;
            report += "  Order #" + order.id;
            report += " - Date: " + order.date;
            report += " - Amount: $" + order.total + "\n";

            for (const item of order.items) {
                report += "    - " + item.productName;
                report += " (" + item.quantity + ")";
                report += " - $" + item.price + "\n";
            }
        }

        report += "\nSummary:\n";
        report += "Total Orders: " + orderCount + "\n";
        report += "Total Spent: $" + totalSpent + "\n";

        if (totalSpent > 1000) {
            report += "VIP Customer - High Value!\n";
        }

        return report;
    }

    updateCustomerLoyaltyPoints(customer) {
        let currentPoints = customer.loyaltyPoints;
        let newPoints = 0;

        const lastMonth = new Date();
        lastMonth.setMonth(lastMonth.getMonth() - 1);

        for (const order of customer.orders) {
            if (new Date(order.date) > lastMonth) {
                newPoints += Math.floor(order.total / 10);
            }
        }

        customer.loyaltyPoints = currentPoints + newPoints;

        if (customer.loyaltyPoints > 1000 && !customer.premium) {
            customer.premium = true;
            customer.memberSince = new Date();
            this.sendPremiumUpgradeEmail(customer.email);
        }
    }

    // Helper methods
    getDatabase() { return null; }
    getInventory() { return null; }
    getEmailService() { return null; }
    sendPremiumUpgradeEmail(email) {}
    buildConfirmationEmail(order) { return ""; }
}

module.exports = { CodeSmellExamples };

File: samples cs/LongMethodFeatureEnvy.cs
using System;
using System.Collections.Generic;
using System.Text;

public class CodeSmellExamples
{
    public void ProcessOrder(Order order)
    {
        // Validation
        if (order == null)
        {
            Console.Error.WriteLine("Order is null");
            return;
        }
        if (order.Items == null || order.Items.Count == 0)
        {
            Console.Error.WriteLine("Order has no items");
            return;
        }

        // Calculate totals
        double subtotal = 0.0;
        foreach (var item in order.Items)
        {
            if (item.Quantity <= 0)
            {
                Console.Error.WriteLine("Invalid quantity for item: " + item.ProductId);
                continue;
            }
            double itemTotal = item.Price * item.Quantity;
            subtotal += itemTotal;
        }

        // Apply discounts
        double discount = 0.0;
        if (order.Customer.IsPremium)
        {
            discount = subtotal * 0.1;
        }
        if (subtotal > 100)
        {
            discount += 10.0;
        }

        // Calculate tax
        double taxRate = 0.08;
        string state = order.Customer.Address.State;
        if (state == "CA")
        {
            taxRate = 0.0725;
        }
        else if (state == "NY")
        {
            taxRate = 0.04;
        }
        else if (state == "TX")
        {
            taxRate = 0.0625;
        }
        double tax = (subtotal - discount) * taxRate;

        // Final total
        double total = subtotal - discount + tax;
        order.Total = total;

        // Update inventory
        foreach (var item in order.Items)
        {
            Inventory inventory = GetInventory();
            Product product = inventory.FindProduct(item.ProductId);
            if (product != null)
            {
                int newQuantity = product.Quantity - item.Quantity;
                product.Quantity = newQuantity;
                inventory.UpdateProduct(product);
            }
        }

        // Save to database
        try
        {
            Database db = GetDatabase();
            db.Save(order);
            Console.WriteLine("Order saved: " + order.Id);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine("Failed to save order: " + e.Message);
            throw new InvalidOperationException("Order processing failed");
        }

        // Send confirmation
        EmailService emailService = GetEmailService();
        string customerEmail = order.Customer.Email;
        string message = BuildConfirmationEmail(order);
        emailService.SendEmail(customerEmail, "Order Confirmation", message);

        // Log activity
        AppLogger logger = GetLogger();
        logger.Log("Order processed: " + order.Id);
        logger.Log("Customer: " + order.Customer.Name);
        logger.Log("Total: $" + total);
    }

    public string GenerateCustomerReport(Customer customer)
    {
        var report = new StringBuilder();
        report.AppendLine("Customer Report");
        report.AppendLine("================\n");

        report.Append("Name: ").Append(customer.FirstName)
              .Append(" ").AppendLine(customer.LastName);
        report.Append("Email: ").AppendLine(customer.Email);
        report.Append("Phone: ").AppendLine(customer.Phone);

        Address address = customer.Address;
        report.Append("Address: ").Append(address.Street).Append(", ");
        report.Append(address.City).Append(", ");
        report.Append(address.State).Append(" ");
        report.AppendLine(address.ZipCode);

        report.Append("Account Status: ");
        if (customer.IsPremium)
        {
            report.AppendLine("Premium Member");
            report.Append("Member Since: ").AppendLine(customer.MemberSince.ToString());
        }
        else
        {
            report.AppendLine("Standard Member");
        }

        report.AppendLine("\nOrder History:");
        int orderCount = 0;
        double totalSpent = 0.0;
        foreach (var order in customer.Orders)
        {
            orderCount++;
            totalSpent += order.Total;
            report.Append("  Order #").Append(order.Id);
            report.Append(" - Date: ").Append(order.Date);
            report.Append(" - Amount: $").AppendLine(order.Total.ToString());

            foreach (var item in order.Items)
            {
                report.Append("    - ").Append(item.ProductName);
                report.Append(" (").Append(item.Quantity).Append(")");
                report.Append(" - $").AppendLine(item.Price.ToString());
            }
        }

        report.AppendLine("\nSummary:");
        report.Append("Total Orders: ").AppendLine(orderCount.ToString());
        report.Append("Total Spent: $").AppendLine(totalSpent.ToString());

        if (totalSpent > 1000)
        {
            report.AppendLine("VIP Customer - High Value!");
        }

        return report.ToString();
    }

    public void UpdateCustomerLoyaltyPoints(Customer customer)
    {
        int currentPoints = customer.LoyaltyPoints;
        int newPoints = 0;

        foreach (var order in customer.Orders)
        {
            if (order.Date > GetLastMonthDate())
            {
                newPoints += (int)(order.Total / 10);
            }
        }

        customer.LoyaltyPoints = currentPoints + newPoints;

        if (customer.LoyaltyPoints > 1000 && !customer.IsPremium)
        {
            customer.IsPremium = true;
            customer.MemberSince = DateTime.Now;
            SendPremiumUpgradeEmail(customer.Email);
        }
    }

    // Helper methods
    private Database GetDatabase() { return null; }
    private Inventory GetInventory() { return null; }
    private EmailService GetEmailService() { return null; }
    private AppLogger GetLogger() { return null; }
    private DateTime GetLastMonthDate() { return DateTime.Now.AddMonths(-1); }
    private void SendPremiumUpgradeEmail(string email) { }
    private string BuildConfirmationEmail(Order order) { return ""; }
}

// Supporting classes
class Customer
{
    public string Id { get; set; }
    public string FirstName { get; set; }
    public string LastName { get; set; }
    public string Name => FirstName + " " + LastName;
    public string Email { get; set; }
    public string Phone { get; set; }
    public Address Address { get; set; }
    public bool IsPremium { get; set; }
    public DateTime MemberSince { get; set; }
    public int LoyaltyPoints { get; set; }
    public List<Order> Orders { get; set; } = new List<Order>();
}

class Address
{
    public string Street { get; set; }
    public string City { get; set; }
    public string State { get; set; }
    public string ZipCode { get; set; }
}

class Order
{
    public string Id { get; set; }
    public Customer Customer { get; set; }
    public List<OrderItem> Items { get; set; } = new List<OrderItem>();
    public double Total { get; set; }
    public DateTime Date { get; set; }
}

class OrderItem
{
    public string ProductId { get; set; }
    public string ProductName { get; set; }
    public int Quantity { get; set; }
    public double Price { get; set; }
}

class Product
{
    public string Id { get; set; }
    public int Quantity { get; set; }
}

class Database
{
    public void Save(Order order) { }
}

class Inventory
{
    public Product FindProduct(string id) { return null; }
    public void UpdateProduct(Product product) { }
}

class EmailService
{
    public void SendEmail(string to, string subject, string message) { }
}

class AppLogger
{
    public void Log(string message) { }
}
